### ISSUE TITLE
build own detect trainer

### ISSUE THREAD

**Issue by bighan123** [ultralytics/ultralytics#22451]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

As the titile mentioned, I want to build my detect trainer by inheriting from the BaseTrainer or DetectTrainer (the base class of yolo), in which I can use the distill train strategy to refine my model, is their any docs can give some advices or helps?  looking for your reply and thanks very much

**Comment by UltralyticsAssistant** [comment]
👋 Hello @bighan123, thank you for your interest in Ultralytics 🚀! This is an automated assistant response — an Ultralytics engineer will review and follow up soon.

For getting started and exploring customization options, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. These resources often cover foundations you can adapt when extending trainers or experimenting with advanced strategies like distillation.

If this is a custom training ❓ Question (e.g., building a custom Detect trainer or integrating distillation), please share:
- A brief outline of what you’ve tried so far (e.g., subclassing approach, hooks used, where you inject teacher/student logic)
- Minimal code snippets showing your custom trainer setup
- Training logs and any console output
- Dataset details or small, representative samples
- Confirmation you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)
These details will help us provide targeted guidance faster 🙌

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Great question—yes, you can do this by subclassing `DetectionTrainer` and returning a `DetectionModel` that overrides `init_criterion()` to wrap the default loss with your KD term; see the Advanced Customization guide and API refs for details: [Advanced Customization (Base/DetectionTrainer)](https://docs.ultralytics.com/usage/engine/), [DetectionTrainer reference](https://docs.ultralytics.com/reference/models/yolo/detect/train/), and [train() with a custom trainer](https://docs.ultralytics.com/reference/engine/model/).

```python
# minimal KD skeleton
from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.nn.tasks import DetectionModel
import torch
import torch.nn as nn

class DistillLoss(nn.Module):
    def __init__(self, teacher, base_criterion, alpha=0.3, T=2.0):
        super().__init__()
        self.teacher = teacher.eval()
        for p in self.teacher.parameters():
            p.requires_grad = False
        self.base, self.alpha, self.T = base_criterion, alpha, T

    def forward(self, preds, batch):
        loss, items = self.base(preds, batch)  # standard YOLO loss
        with torch.no_grad():
            t_preds = self.teacher(batch["img"])  # teacher forward
        # TODO: add KD term(s) matching your head outputs, e.g. KL on cls logits, L1 on boxes
        kd = torch.zeros(1, device=loss.device)
        loss = loss + self.alpha * kd
        items["kd"] = kd
        return loss, items

class KDModel(DetectionModel):
    def init_criterion(self):
        base = super().init_criterion()               # default YOLO loss
        teacher = YOLO("yolo11m.pt").model            # choose any stronger teacher
        self.criterion = DistillLoss(teacher, base)
        return self.criterion

class KDTrainer(DetectionTrainer):
    def get_model(self, cfg, weights, verbose: bool = True):
        m = KDModel(cfg, nc=self.data["nc"], ch=self.data["channels"], verbose=verbose)
        if weights:
            m.load(weights)
        return m

# usage
student = YOLO("yolo11n.yaml")
results = student.train(trainer=KDTrainer, data="coco8.yaml", epochs=3)
```

If you prefer callbacks to keep the model untouched, the same page shows where to plug into training events; start there and iterate.

**Comment by bighan123** [comment]
Thank you very much!  I follow the advice and create my own Detect Trainer.  but I still have a question. I note that within the native `_do_train` function, `self.model(batch)` returns the loss. Now I require access to the model's predicted features and logits. I have achieved this by modifying the internal forward function of self.model, it looks like:
''' 
(self.loss, self.loss_items), student_preds = self.model(batch, return_preds=True)
'''

However, I wish to avoid disrupting the native training pipeline. I would like to know whether it is possible to obtain the model's predicted features by adding a hook function, without explicitly modifying self.model(batch).

**Comment by glenn-jocher** [comment]
Yes—attach a PyTorch forward hook to the Detect head (last module) to cache raw preds during the normal training forward, then read them in a callback; no need to change `self.model(batch)`.

```python
from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.nn.tasks import DetectionModel

class HookedTrainer(DetectionTrainer):
    def get_model(self, cfg, weights, verbose=True):
        m = DetectionModel(cfg, nc=self.data["nc"], ch=self.data["channels"], verbose=verbose)
        if weights:
            m.load(weights)
        m._cache = {}

        def save_head(_, __, out):
            out = out if isinstance(out, (list, tuple)) else [out]
            m._cache["preds"] = [o.detach() for o in out]  # detach to avoid graph retention

        m.model[-1].register_forward_hook(save_head)  # Detect head is last module
        return m

def on_train_batch_end(trainer):
    preds = trainer.model._cache.pop("preds", None)  # tensors per scale: logits/boxes before NMS
    if preds is not None:
        pass  # use or log preds here

model = YOLO("yolo11n.yaml")
model.add_callback("on_train_batch_end", on_train_batch_end)
model.train(trainer=HookedTrainer, data="coco8.yaml", epochs=3)
```

For where to plug your logic, see the callback events in the [Callbacks guide](https://docs.ultralytics.com/usage/callbacks/), and for why the head is addressable via `model.model[-1]`, see the [DetectionModel reference](https://docs.ultralytics.com/reference/nn/tasks/).
